### ISSUE TITLE
'nodeIntegrationInWorker' no longer affects worklets, enabling node integration and 'require' impossible? (regression)

### ISSUE THREAD

**Issue by JohnWeisz** [electron/electron#22503]
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:** ~~8.0.3~~ 7.0.0 to 11.x
* **Operating System:** Windows 10 (1909)
* **Last Known Working Electron version:** ~~4.0.0-beta.4~~ 6.1.9

### Expected Behavior

Given a worklet thread, such as an AudioWorklet thread (AudioWorkletGlobalScope), spawned from a BrowserWindow main thread that has both `nodeIntegration` and `nodeIntegrationInWorker` enabled, `require` should work exactly as it does in regular web workers (i.e. DedicatedWorkerGlobalScope) and node integration should be generally available.

### Actual Behavior

Unlike in regular worker threads, node integration no longer gets enabled in worklet threads, and calling require results in the following error message:

```
'require' is not defined.
```

### To Reproduce

Reproducing the issue itself is very simple, but it involves some setup.

Spawn a BrowserWindow process with `nodeIntegration` and `nodeIntegrationInWorker` enabled, and from within, inside an AudioWorklet, attempt to call `require`. This will result in an error, as node integration fails to become enabled in AudioWorklet threads (I assume every worklet thread is affected).

I understand this may not be trivial to setup, and as such I'll try to put together a complete example if I have the time. For the time being, reproducing the issue itself is as simple as calling `require` from an AudioWorklet thread, which (unlike a `require` call from a regular web worker thread) will throw due to being missing.

### Additional Information

The current workaround is to revert to an older version of Electron, which is a solid _confirmed_ fix. As of yet I'm unsure when this broke, but I know for certain that it used to work in `4.0.0-beta.4`. I'll keep this issue updated.

I also know for certain that the availability of `require` (and co) in worklets used to be influenced by the `nodeIntegrationInWorker` setting before, identical to regular workers.

**Edit:** this still works in `6.0.0`

**Comment by JohnWeisz** [comment]
**Update:**

 - works in `6.1.9`
 - does NOT work in `7.0.0`

Since `6.1.9` was the latest stable release in major version 6, I assume this is the last point at which this worked, before getting broken in major version 7.

**Comment by JohnWeisz** [comment]
On a personal thought, I always felt like the `nodeIntegrationInWorkers` setting affecting worklet threads was coincidental. Perhaps introducing a `nodeIntegrationInWorklets` setting separately would be a sound idea while we are at it?

**Comment by NahejL** [comment]
If it can help, running an app with `nodeIntegrationInWorker: true,` and using an AudioWorklet results in: 

**terminal:**
> [3860:0909/024426.701283:ERROR:CONSOLE(116)] "Request Network.enable failed. {"code":-32601,"message":"'Network.enable' wasn't found"}", source: devtools://devtools/bundled/protocol_client/protocol_client.js (116)

**webTools console:**
> Uncaught ReferenceError: self is not defined
>     at Object../lib/renderer/webpack-provider.ts (electron/js2c/worker_init.js:97)
>     at __webpack_require__ (electron/js2c/worker_init.js:1)
>     at Object../lib/worker/init.js (electron/js2c/worker_init.js:101)
>     at __webpack_require__ (electron/js2c/worker_init.js:1)
>     at electron/js2c/worker_init.js:1
>     at electron/js2c/worker_init.js:1
>     at NativeModule.compileForInternalLoader (internal/bootstrap/loaders.js:279)
>     at NativeModule.compileForPublicLoader (internal/bootstrap/loaders.js:219)
>     at loadNativeModule (internal/modules/cjs/helpers.js:26)
>     at Module._load (internal/modules/cjs/loader.js:852)

**also, after reload**
no errors are logged (main nor renderer) and the html script is "half executed" 

sorry for bad comment formatting, never commented before but I'd like to fix this issue

**Comment by JohnWeisz** [comment]
Still broken in `10.1.3`, which now forces worklet node integration to run 4 major versions behind latest. Anything we can do to help move this forward?

**Comment by jsonfry** [comment]
Can confirm still broken in `10.1.5`, running on Linux (Ubuntu 20.04). Seeing the same logs as above https://github.com/electron/electron/issues/22503#issuecomment-689346726

**Comment by rodw** [comment]
Bump. Also seeing output like https://github.com/electron/electron/issues/22503#issuecomment-689346726 in Electron 11.1.1.

Is there any news on this front?

**Comment by pl4yradam** [comment]
I hope this helps anyone:

I'm using electron **11.1.1** and had the same issue.
Solution on StackOverflow: https://stackoverflow.com/questions/52760219/use-audioworklet-within-electron-domexception-the-user-aborted-a-request

**My code interpretation:**
`                   const processorPath = <YOUR_FILE>;
                    const processorSource = fs.readFileSync(processorPath);
                    const processorBlob = new Blob([processorSource.toString()], { type: 'text/javascript' });
                    const processorURL = URL.createObjectURL(processorBlob);

                    await this.audioContext.audioWorklet.addModule(processorURL);`

**Comment by JohnWeisz** [comment]
Thanks for that @pl4yradam but the problem is doing a require call _inside_ the AudioWorklet module you are loading in your code snippet, i.e. the `processorPath` module. Your code snippet runs on the main thread, which works just fine in this regard.

As for the AudioWorklet module itself, well of course one solution is doing cross-thread audio processing with atomics and using a dedicated Worker to do whatever you need, while the AudioWorklet just sends and receives data from that Worker thread (usually through a ring buffer).

**Comment by hamst** [comment]
There is an issue with fixing the problem: 
- An init script inside electron is used to initialize node environment for worker thread (lib/worker/init.js or lib/worker/init.ts).
- The script is used as well for any type of worklets, including AudioWorklet.
- The script currently relies on `location` symbol of `WorkerGlobalScope` for proper module initialization. The path to js file with worker implementation from `location` allows to use __dirname and make `require` able to import modules by relative path.
- `location` symbol is not exposed in any worklet scope (e.g. AudioWorkletGlobalScope), so there is no way to get the actual path of the module with worklet implementation and make `require` work as expected.

As a workaround that could be done, I proposed adding `preloadInWorker` option: the preload script can use node API without any limitation and define some functions with `require` inside to be used in the worklet scope. (see details https://github.com/electron/electron/issues/28620).

**Comment by gpetrov** [comment]
still an issue in Electron 12.0.5, hopefully will be fixed soon. it happens when loading css worklets

**Comment by nornagon** [comment]
Would it be possible to make a standalone testcase with only the code necessary to reproduce the issue? For example, [Electron Fiddle](https://www.electronjs.org/fiddle) is a great tool for making small test cases and makes it easy to publish your test case to a [gist](https://gist.github.com) that Electron maintainers can use.

Stand-alone test cases make fixing issues go more smoothly: it ensure everyone's looking at the same issue, it removes all unnecessary variables from the equation, and it can also provide the basis for automated regression tests.

I'm adding the `blocked/need-repro` label for this reason.

**Comment by gpetrov** [comment]
@nornagon  sure here is a repro based on electron fiddle.

[worklet_bug.zip](https://github.com/electron/electron/files/6387928/worklet_bug.zip)


The errors happen when `nodeIntegrationInWorker: true` and a worklet is used. Doesn't seem to effect to working of the worklet, just weird js errors pop in console:

```
electron/js2c/worker_init.js:121 Electron worker_init.js script failed to run
(anonymous) @ electron/js2c/worker_init.js:121
electron/js2c/worker_init.js:121 ReferenceError: self is not defined
    at Object.<anonymous> (electron/js2c/worker_init.js:97)
    at Object../lib/worker/init.ts (electron/js2c/worker_init.js:97)
    at __webpack_require__ (electron/js2c/worker_init.js:1)
    at electron/js2c/worker_init.js:1
    at electron/js2c/worker_init.js:1
    at NativeModule.compileForInternalLoader (internal/bootstrap/loaders.js:283)
    at NativeModule.compileForPublicLoader (internal/bootstrap/loaders.js:225)
    at loadNativeModule (internal/modules/cjs/helpers.js:35)
    at Module._load (internal/modules/cjs/loader.js:747)
    at Function.f._load (electron/js2c/asar_bundle.js:5)
```

**Comment by marcelblum** [comment]
Testing this today in an Electron 12.0.16 project and just want to clarify from earlier comments, **Node integration does seem to work fine in the worklet despite the console errors**. `require` does work it just needs an absolute path when run from within the worklet. What I've been doing is creating a `Blob` URL of my worklet code in the renderer main thread with dynamically inserted absolute paths to needed modules using `require.resolve()` so that it works in production/packaged form. Along the lines of:

```
const codeBlobURL = URL.createObjectURL(new Blob([`
 class SomeProcessor extends AudioWorkletProcessor {
  process (inputs, outputs, parameters) {
   //...
   return true
  }
 }
 const someNodeModuleInstance = require(${JSON.stringify(require.resolve('someNodeModule'))});
 //...
 registerProcessor('some-processor', SomeProcessor )
`], { type: "application/javascript; charset=utf-8" }))
```

**Comment by rodw** [comment]
For what it's worth my experience with AudioWorklets in Electron v13 matches that described by @marcelblum.

I'm currently using Electron v13.1.7 but I've seen the same behavior for a while. IIRC it started with Electron v7 but was not an issue in v6.

I've got `nodeIntegrationInWorker` set to `true` and I'm using several AudioWorkletProcessor implementations that are bootstrapped more or less like Marcel's example (loading the raw source code for the worklet as a blob and taking pains to ensure each AudioWorkletProcessor-implementing class is only registered (via `registerProcessor`) once as I'm pretty sure I ran into problems without that).

I'm able to `require` external modules (package.json dependencies) in my worklets without any problem. I'm not currently loading any local files - in part because I had trouble getting that to work consistently for packaged and un-packaged  apps, but the absolute-path topic probably explains that. (To be clear, as it happens some of my worklets are completely dependency-free - no `require` calls at all -  and exhibit the same symptoms, so I don't think that `require` is a direct trigger.)

Each time registerProcessor is invoked for a new type of AudioWorkletProcessor a pair of error messages is logged:

```
electron/js2c/worker_init.js:121 Electron worker_init.js script failed to run
(anonymous) @ electron/js2c/worker_init.js:121

electron/js2c/worker_init.js:121 ReferenceError: self is not defined
    at Object.<anonymous> (electron/js2c/worker_init.js:97)
    at Object../lib/worker/init.ts (electron/js2c/worker_init.js:97)
    at __webpack_require__ (electron/js2c/worker_init.js:1)
    at electron/js2c/worker_init.js:1
    at electron/js2c/worker_init.js:1
    at NativeModule.compileForInternalLoader (internal/bootstrap/loaders.js:283)
    at NativeModule.compileForPublicLoader (internal/bootstrap/loaders.js:225)
    at loadNativeModule (internal/modules/cjs/helpers.js:35)
    at Module._load (internal/modules/cjs/loader.js:747)
    at Function.f._load (electron/js2c/asar_bundle.js:5)
(anonymous) @ electron/js2c/worker_init.js:121
```

The exact error message may have changed slightly from release to release of Electron, but it's otherwise the exact same message and trace every time.

This happens once for each AudioWorkletProcessor implementation I invoke.

I don't *think* this causes any other specific problem. I.e., as far as I know there are no side-effects other than these error messages appearing in the log. If that's true this is more of an annoyance than a problem. 

But the error itself is a little a confidence-shaking. Worklets seem a little fussy in general so (for me at least) it's a little hard to tell if quirks I encounter in Electron (like needing to load the worklet as Blob via createObjectURL as a Blob) are working as intended or if they point to some underlying limitations or problems.

Is there any way (anyone one) to confirm that this is *just* a spurious error message that one can more-or-less ignore?

**Comment by arcusfelis** [comment]
The error is annoying, would rather Electron would not try to load node integration at all in worklets, if it is broken.

```
node:electron/js2c/worker_init:97 Electron worker_init.js script failed to run
(anonymous) @ node:electron/js2c/worker_init:97
node:electron/js2c/worker_init:97 ReferenceError: self is not defined
    at Object.<anonymous> (node:electron/js2c/worker_init:73:565)
    at ./lib/worker/init.ts (node:electron/js2c/worker_init:73:1093)
    at __webpack_require__ (node:electron/js2c/worker_init:1:132)
    at node:electron/js2c/worker_init:1:1204
    at node:electron/js2c/worker_init:1:1270
    at NativeModule.compileForInternalLoader (node:internal/bootstrap/loaders:312:7)
    at NativeModule.compileForPublicLoader (node:internal/bootstrap/loaders:252:10)
    at loadNativeModule (node:internal/modules/cjs/helpers:49:9)
    at Module._load (node:internal/modules/cjs/loader:811:15)
    at c._load (node:electron/js2c/asar_bundle:5:13343)
```

**Comment by marcelblum** [comment]
@arcusfelis `nodeIntegrationInWorker` is not broken. That's just an apparently inconsequential error message that appears in the console. `require` in AudioWorklets still seems to be working fine as of latest Electron 20.1.4.

**Comment by JohnWeisz** [comment]
I'd say let's bump this.

While it sounds like an in-thread workaround is possible (and also, you can just fire up a regular worker thread, then communicate with it from the worklet using a MessageChannel), this is still borderline broken.

**Comment by rodw** [comment]
@JohnWeisz can you clarify what problem the in-thread workaround is working around?

I haven't verified this with the most recent Electron releases, but reading this thread my understanding is that [marcelblum's summary above](https://github.com/electron/electron/issues/22503#issuecomment-897100531) was and continues to be accurate. Namely, that worklets with or without node dependencies are working as designed (assuming you get your `require` paths right), save for a confidence-shaking error message and stack trace being sent to console.error a finite number of times (roughly once per worklet registration maybe?)

Is that accurate?

Assuming yes, I don't exactly disagree with your "borderline broken" assessment. I'm just trying to confirm that I understand the observed behavior and how it is different from the expected behavior.

**Comment by marcelblum** [comment]
Looks like this was fixed in the latest electron releases via #37041. Error message now gone, at least in my project using v.22.2.1 on Windows.

**Comment by marcelblum** [comment]
To qualify my above comment, the errors in the node console emitted from the main process are still present ("Request Network.enable failed" etc). Errors in the renderer are gone ("self is not defined" etc.)
